File: DataStructures/src/main/java/com/zhangdd/tree/BinaryTree.java
package com.zhangdd.tree;

/**
 * @author zhangdd on 2021/11/11
 */
public class BinaryTree {

    private TreeNode root;

    public BinaryTree(TreeNode root) {
        this.root = root;
    }

    public void preOrder() {
        if (null == root) {
            System.out.println("当前树空");
            return;
        }
        root.preOrder();
    }

    public void infixOrder() {
        if (null == root) {
            System.out.println("当前树空");
            return;
        }
        root.infixOrder();
    }

    public void postOrder() {
        if (null == root) {
            System.out.println("当前树空");
            return;
        }
        root.postOrder();
    }

    public TreeNode preSearch(int no) {
        if (null == root) {
            System.out.println("当前树空");
            return null;
        }
        return root.preSearch(no);
    }

    public TreeNode infixSearch(int no) {
        if (null == root) {
            System.out.println("当前树空");
            return null;
        }

        return root.infixSearch(no);
    }

    public TreeNode postSearch(int no) {
        if (null == root) {
            System.out.println("当前树空");
            return null;
        }

        return root.postSearch(no);
    }

    public void remove(int no) {
        if (null == root) {
            System.out.println("当前树空");
            return;
        }
        if (root.no == no) {
            root = null;
        }
        root.remove(no);
    }


    public static class TreeNode {
        private int no;
        private String name;
        private TreeNode left;
        private TreeNode right;

        public TreeNode(int no, String name) {
            this.no = no;
            this.name = name;
        }

        public int getNo() {
            return no;
        }

        public void setNo(int no) {
            this.no = no;
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public TreeNode getLeft() {
            return left;
        }

        public void setLeft(TreeNode left) {
            this.left = left;
        }

        public TreeNode getRight() {
            return right;
        }

        public void setRight(TreeNode right) {
            this.right = right;
        }

        public void preOrder() {
            System.out.println(this);
            if (null != this.left) {
                this.left.preOrder();
            }
            if (null != this.right) {
                this.right.preOrder();
            }
        }

        public void infixOrder() {
            if (null != this.left) {
                this.left.infixOrder();
            }
            System.out.println(this);
            if (null != this.right) {
                this.right.infixOrder();
            }
        }

        public void postOrder() {
            if (null != this.left) {
                this.left.postOrder();
            }
            if (null != this.right) {
                this.right.postOrder();
            }
            System.out.println(this);
        }

        public TreeNode preSearch(int no) {
            if (this.no == no) {
                return this;
            }
            TreeNode result = null;
            if (null != this.left) {
                result = this.left.preSearch(no);
            }
            if (null != result) {
                return result;
            }
            if (null != this.right) {
                result = this.right.preSearch(no);
            }

            return result;
        }

        public TreeNode infixSearch(int no) {
            TreeNode result = null;
            if (null != this.left) {
                result = this.left.infixSearch(no);
            }
            if (null != result) {
                return result;
            }
            if (this.no == no) {
                return this;
            }
            if (null != this.right) {
                result = this.right.infixSearch(no);
            }

            return result;
        }

        public TreeNode postSearch(int no) {
            TreeNode result = null;
            if (null != this.left) {
                result = this.left.postSearch(no);
            }
            if (null != result) {
                return result;
            }

            if (null != this.right) {
                result = this.right.postSearch(no);
            }

            if (null != result) {
                return result;
            }

            if (this.no == no) {
                return this;
            }

            return result;
        }

        public void remove(int no) {
            if (null != this.left && this.left.no == no) {
                this.left = null;
                return;
            }
            if (null != this.right && this.right.no == no) {
                this.right = null;
                return;
            }
            if (null != this.left) {
                this.left.remove(no);
            }
            if (null != this.right) {
                this.right.remove(no);
            }
        }

        @Override
        public String toString() {
            return "TreeNode{" +
                    "no=" + no +
                    ", name='" + name + '\'' +
                    '}';
        }
    }
}

File: python/DataStructures/tree/BinaryTree.py
#!/usr/bin/env python3
"""
@author: zhangdd on 2022/2/4
"""


class Node:
    def __init__(self, no, name):
        self.no = no
        self.name = name
        self.left = None
        self.right = None

    def show(self):
        print('no:{},name:{}'.format(self.no, self.name))

    def pre_order(self):
        self.show()
        if self.left is not None:
            self.left.pre_order()
        if self.right is not None:
            self.right.pre_order()

    def infix_order(self):
        if self.left is not None:
            self.left.infix_order()

        self.show()

        if self.right is not None:
            self.right.infix_order()

    def post_order(self):
        if self.left is not None:
            self.left.post_order()

        if self.right is not None:
            self.right.post_order()

        self.show()

    def remove(self, no):
        if self.left is not None:
            if self.left.no == no:
                self.left = None
            else:
                self.left.remove(no)

        if self.right is not None:
            if self.right.no == no:
                self.right = None
            else:
                self.right.remove(no)

    def set_left(self, node):
        self.left = node

    def set_right(self, node):
        self.right = node


class BinaryTree:
    def __init__(self, node):
        self.root = node

    def pre_order(self):
        if self.root is None:
            print('当前树空')
            return
        self.root.pre_order()

    def infix_order(self):
        if self.root is None:
            print('当前树空')
            return
        self.root.infix_order()

    def post_order(self):
        if self.root is None:
            print('当前树空')
            return
        self.root.post_order()

    def remove(self, no):
        if self.root is None:
            print('当前树空')
            return
        if self.root.no == no:
            self.root = None
            return
        self.root.remove(no)


if __name__ == '__main__':
    root = Node(1, '宋江')
    node2 = Node(2, '吴用')
    node3 = Node(3, '卢俊义')
    node4 = Node(4, '林冲')
    node5 = Node(5, '关胜')

    root.set_left(node2)
    root.set_right(node3)

    node3.set_left(node5)
    node3.set_right(node4)

    tree = BinaryTree(root)

    print('前序')
    tree.pre_order()

    print('中序')
    tree.infix_order()

    print('后序')
    tree.post_order()

    print('-' * 50)
    tree.remove(4)
    print('删除节点4后')
    tree.pre_order()
    print('-' * 50)

    tree.remove(1)
    print('删除节点1后')
    tree.pre_order()
